presto-main/src/main/java/io/prestosql/operator/WindowOperator.java

```diff
@@ -607,6 +623,68 @@ public TransformationState<Page> process(WindowPartition partition)
         }
     }
 
+    private class LimitWindowPartitionsToOutputPages
+            implements Transformation<WindowPartition, Page>
+    {
+        final PageBuilder tempPageBuilder;
+        final PageBuilder pageBuilder;
+        final LimitWindowFunction function;
+
+        LimitWindowPartitionsToOutputPages(LimitWindowFunction function)
+        {
+            tempPageBuilder = new PageBuilder(outputTypes);
+            pageBuilder = new PageBuilder(outputTypes);
+            this.function = function;
+        }
+
+        @Override
+        public TransformationState<Page> process(WindowPartition partition)
+        {
+            boolean finishing = partition == null;
+            if (finishing) {
+                if (tempPageBuilder.isEmpty() && pageBuilder.isEmpty()) {
+                    return TransformationState.finished();
+                }
+
+                // Output the remaining page if we have anything buffered
+                int[] positions = function.select(tempPageBuilder.getPositionCount());
+                copyPositions(tempPageBuilder, pageBuilder, positions);
+                Page page = pageBuilder.build();
+                pageBuilder.reset();
+                return TransformationState.ofResult(page, false);
+            }
+
+            // process whole partition
+            // TODO: When a single partition is too large, OOM could happen
+            while (partition.hasNext()) {
+                partition.processNextRow(tempPageBuilder);
+            }
+            int[] positions = function.select(tempPageBuilder.getPositionCount());
+            copyPositions(tempPageBuilder, pageBuilder, positions);
+
```

Comment:
Yeah. The memory consumption of tempPageBuilder looks huge for large input tables. Can't we have a reservoir buffer within this transformation class and perform the sampling here?

Reply:
I guess the memory usage difference is:
- current: (WindowPartition size) x (the number of unique users)
- optimum: (reservoir size) x (the number of unique users)

Or tempPageBuilder can be quickly reduced to the reservoir size as Presto processes WindowPartition?

Reply:
Not sure how often we have a partition-size >>> sample-size. 
One idea is a cascading page buffer. For example, tempBuffer1.ifFull -> select().create(tempBuffer2), tempBuffer2.ifFull -> select().create(tempBuffer3) an so on.

Reply:
As we only need to maintain a fixed-size reservoir and the input records up to its size, adding a sampling around partition.processNextRow(...) line looks more natural to me. Isn't it possible?